From ceb763deda6395d5935f73e6d6f13b76de003da1 Mon Sep 17 00:00:00 2001
From: shahed <jamshidalinezhadian@gmail.com>
Date: Wed, 23 Sep 2026 07:30:29 +0330
Subject: [PATCH 1/2] netd: skip cgroup v2 and libnetd_updatable init without
 eBPF

Successor to the Android 12 'Ignore netd errors for < 4.9 kernels'
patch, whose target (trafficCtrl in Controllers.cpp) moved out of netd
in Android 13. Kernels older than 4.9 have no cgroup v2 root and no
eBPF, so netd's main() exited on both the cgroup v2 lookup and
libnetd_updatable_init. When ro.kernel.ebpf.supported is false, skip
both.

Change-Id: Ib40de140a48bd98132285c62b074a21d03902438
---
 server/main.cpp | 28 +++++++++++++++++++---------
 1 file changed, 19 insertions(+), 9 deletions(-)

diff --git a/server/main.cpp b/server/main.cpp
index 0e81d4e..af24f33 100644
--- a/server/main.cpp
+++ b/server/main.cpp
@@ -36,6 +36,7 @@
 #include <binder/IServiceManager.h>
 #include <netdutils/Stopwatch.h>
 #include <processgroup/processgroup.h>
+#include <android-base/properties.h>
 
 #include "Controllers.h"
 #include "FwmarkServer.h"
@@ -126,17 +127,26 @@ int main() {
         gLog.info("setCloseOnExec(%s)", sock);
     }
 
-    std::string cg2_path;
-    if (!CgroupGetControllerPath(CGROUPV2_CONTROLLER_NAME, &cg2_path)) {
-        ALOGE("Failed to find cgroup v2 root %s", strerror(errno));
-        exit(1);
-    }
+    // Kernels older than 4.9 (kiwi's 3.10) have neither cgroup v2 nor eBPF: there
+    // is no cgroup v2 root, and libnetd_updatable_init, which attaches BPF
+    // programs to it, cannot succeed. Android 13 exits netd in both cases. When
+    // ro.kernel.ebpf.supported is false, skip both. Successor to the Android 12
+    // 'Ignore netd errors for < 4.9 kernels' patch, whose target moved out of netd.
+    if (android::base::GetBoolProperty("ro.kernel.ebpf.supported", true)) {
+        std::string cg2_path;
+        if (!CgroupGetControllerPath(CGROUPV2_CONTROLLER_NAME, &cg2_path)) {
+            ALOGE("Failed to find cgroup v2 root %s", strerror(errno));
+            exit(1);
+        }
 
-    if (libnetd_updatable_init(cg2_path.c_str())) {
-        ALOGE("libnetd_updatable_init failed");
-        exit(1);
+        if (libnetd_updatable_init(cg2_path.c_str())) {
+            ALOGE("libnetd_updatable_init failed");
+            exit(1);
+        }
+        gLog.info("libnetd_updatable_init success");
+    } else {
+        gLog.info("eBPF not supported by this kernel, skipping libnetd_updatable_init");
     }
-    gLog.info("libnetd_updatable_init success");
 
     NetlinkManager *nm = NetlinkManager::Instance();
     if (nm == nullptr) {
-- 
2.43.0

